src/modules/cpu/kernel/phase.hpp

```diff
@@ -0,0 +1,965 @@
+        srcPtr2Image = srcPtr2 + batchCount * srcDescPtr->strides.nStride;
+        dstPtrImage = dstPtr + batchCount * dstDescPtr->strides.nStride;
+
+        Rpp32u bufferLength = roi.xywhROI.roiWidth * layoutParams.bufferMultiplier;
+
+        __m256 pMul = _mm256_set1_ps(multiplier);
+
+        Rpp16f *srcPtr1Channel, *srcPtr2Channel, *dstPtrChannel;
+        srcPtr1Channel = srcPtr1Image + (roi.xywhROI.xy.y * srcDescPtr->strides.hStride) + (roi.xywhROI.xy.x * layoutParams.bufferMultiplier);
+        srcPtr2Channel = srcPtr2Image + (roi.xywhROI.xy.y * srcDescPtr->strides.hStride) + (roi.xywhROI.xy.x * layoutParams.bufferMultiplier);
+        dstPtrChannel = dstPtrImage;
+
+        // Phase with fused output-layout toggle (NHWC -> NCHW)
+        if ((srcDescPtr->c == 3) && (srcDescPtr->layout == RpptLayout::NHWC) && (dstDescPtr->layout == RpptLayout::NCHW))
+        {
+            Rpp32u alignedLength = (bufferLength / 24) * 24;
+
+            Rpp16f *srcPtr1Row, *srcPtr2Row, *dstPtrRowR, *dstPtrRowG, *dstPtrRowB;
+            srcPtr1Row = srcPtr1Channel;
+            srcPtr2Row = srcPtr2Channel;
+            dstPtrRowR = dstPtrChannel;
+            dstPtrRowG = dstPtrRowR + dstDescPtr->strides.cStride;
+            dstPtrRowB = dstPtrRowG + dstDescPtr->strides.cStride;
+
+            for(int i = 0; i < roi.xywhROI.roiHeight; i++)
+            {
+                Rpp16f *srcPtr1Temp, *srcPtr2Temp, *dstPtrTempR, *dstPtrTempG, *dstPtrTempB;
+                srcPtr1Temp = srcPtr1Row;
+                srcPtr2Temp = srcPtr2Row;
+                dstPtrTempR = dstPtrRowR;
+                dstPtrTempG = dstPtrRowG;
+                dstPtrTempB = dstPtrRowB;
+
+                int vectorLoopCount = 0;
+                for (; vectorLoopCount < alignedLength; vectorLoopCount += 24)
+                {
+                    Rpp32f srcPtr1Temp_ps[24], srcPtr2Temp_ps[24], dstPtrTemp_ps[24];
+
+                    for(int cnt = 0; cnt < 24; cnt++)
+                    {
+                        *(srcPtr1Temp_ps + cnt) = (Rpp32f) *(srcPtr1Temp + cnt);
+                        *(srcPtr2Temp_ps + cnt) = (Rpp32f) *(srcPtr2Temp + cnt);
+                    }
+
+                    __m256 p1[3], p2[3];
+
+                    rpp_simd_load(rpp_load24_f32pkd3_to_f32pln3_avx, srcPtr1Temp_ps, p1);    // simd loads
+                    rpp_simd_load(rpp_load24_f32pkd3_to_f32pln3_avx, srcPtr2Temp_ps, p2);    // simd loads
+                    p1[0] = _mm256_mul_ps(atan2_ps(p1[0], p2[0]), pMul);    // phase computation
+                    p1[1] = _mm256_mul_ps(atan2_ps(p1[1], p2[1]), pMul);    // phase computation
+                    p1[2] = _mm256_mul_ps(atan2_ps(p1[2], p2[2]), pMul);    // phase computation
+                    rpp_simd_store(rpp_store24_f32pln3_to_f32pln3_avx, dstPtrTemp_ps, dstPtrTemp_ps + 8, dstPtrTemp_ps + 16, p1);    // simd stores
+
+                    for(int cnt = 0; cnt < 8; cnt++)
```

Comment:
Instead of manually typecasting and store in output pixel by pixel
try to use _mm256_cvtps_ph followed by _mm_storeu_si128

Reference in rocAL as below
https://github.com/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/blob/master/rocAL/rocAL/source/pipeline/master_graph.cpp#L881

check the performance and mismatches and retain the accurate and performant version

Reply:
Done, yes seeing slight improvements with this.